File: swift/day7_part2_2023.swift
import Foundation

struct Hand {
    var cards: String
    var bid: Int
}

let fileURL = URL(fileURLWithPath: "input.txt")
let input = try String(contentsOf: fileURL)
let lines = input.components(separatedBy: "\n")

var hands = [Hand]()

let re = try NSRegularExpression(pattern: "[\\dAKQJT]+")
let bidRe = try NSRegularExpression(pattern: " [\\d]+")

for line in lines {
    if line.isEmpty {
        continue
    }

    let cardsRange = NSRange(location: 0, length: line.utf16.count)
    let cardsMatch = re.firstMatch(in: line, range: cardsRange)
    let bidMatch = bidRe.firstMatch(in: line, range: cardsRange)

    guard let cardsRange = cardsMatch?.range, let bidRange = bidMatch?.range else { continue }

    let cards = (line as NSString).substring(with: cardsRange)
    let bidString = (line as NSString).substring(with: bidRange).trimmingCharacters(in: .whitespaces)
    guard let bid = Int(bidString) else { continue }

    hands.append(Hand(cards: cards, bid: bid))
}

var matches = [[Hand](), [Hand](), [Hand](), [Hand](), [Hand](), [Hand](), [Hand]()]

let valueDict: [Character: Int] = ["J": 1, "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, "9": 9, "T": 10, "Q": 11, "K": 12, "A": 13]

for hand in hands {
    var count = [Character: Int]()

    for card in hand.cards {
        count[card, default: 0] += 1
    }

    if let jCount = count["J"], jCount > 0 {
        var highV = 0
        var highKey: Character = "J"
        for (key, value) in count {
            if key != "J" {
                if value > highV {
                    highKey = key
                    highV = value
                } else if value == highV && valueDict[key]! > valueDict[highKey]! {
                    highKey = key
                }
            }
        }
        if highKey != "J" {
            count[highKey, default: 0] += jCount
            count["J"] = nil
        }
    }

    var value = 1
    for (_, i) in count {
        value *= i
    }

    switch value {
    case 1:
        matches[6].append(hand)
    case 2:
        matches[5].append(hand)
    case 3:
        matches[3].append(hand)
    case 4:
        if count.count == 2 {
            matches[1].append(hand)
        } else {
            matches[4].append(hand)
        }
    case 5:
        matches[0].append(hand)
    case 6:
        matches[2].append(hand)
    default:
        print("oh no")
    }
}

var convertedMatches = [[Int]]()

for x in matches {
    var temp = [[Int]]()
    for i in x {
        var y = i.cards.replacingOccurrences(of: "A", with: "E")
        y = y.replacingOccurrences(of: "T", with: "A")
        y = y.replacingOccurrences(of: "J", with: "1")
        y = y.replacingOccurrences(of: "Q", with: "C")
        y = y.replacingOccurrences(of: "K", with: "D")
        if let val = Int(y, radix: 16) {
            temp.append([val, i.bid])
        }
    }
    temp.sort { $0[0] > $1[0] }
    convertedMatches.append(contentsOf: temp)
}

var total = 0
for x in 0..<convertedMatches.count {
    total += convertedMatches[x][1] * (convertedMatches.count - x)
}

print(total)

File: dart/day7_part2_2023.dart
import 'dart:io';
import 'dart:convert';
import 'dart:math';

class Hand {
  String cards;
  int bid;

  Hand(this.cards, this.bid);
}

final valueDict = {'J': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9, 'T': 10, 'Q': 11, 'K': 12, 'A': 13};

void main() {
  final file = File('input.txt');
  final input = file.readAsStringSync();
  final lines = input.split('\n');

  final hands = <Hand>[];

  for (final line in lines) {
    if (line.isEmpty) continue;
    final cardMatch = RegExp(r'[\dAKQJT]+').firstMatch(line);
    final cards = cardMatch?.group(0) ?? '';
    final bidMatch = RegExp(r' [\d]+').firstMatch(line);
    final bid = int.parse(bidMatch?.group(0)?.substring(1) ?? '0');
    hands.add(Hand(cards, bid));
  }

  final matches = [[], [], [], [], [], [], []];

  for (final hand in hands) {
    final count = <String, int>{};
    for (final char in hand.cards.split('')) {
      count[char] = (count[char] ?? 0) + 1;
    }

    if (count['J'] != null && count['J']! > 0) {
      int highV = 0;
      String highKey = 'J';
      for (final key in count.keys) {
        if (key != 'J') {
          if (count[key]! > highV) {
            highKey = key;
            highV = count[key]!;
          } else if (count[key]! == highV && valueDict[key]! > valueDict[highKey]!) {
            highKey = key;
          }
        }
      }
      if (highKey != 'J') {
        count[highKey] = (count[highKey] ?? 0) + count['J']!;
        count.remove('J');
      }
    }

    int value = 1;
    for (final i in count.values) {
      value *= i;
    }

    switch (value) {
      case 1:
        matches[6].add(hand);
        break;
      case 2:
        matches[5].add(hand);
        break;
      case 3:
        matches[3].add(hand);
        break;
      case 4:
        if (count.length == 2) {
          matches[1].add(hand);
        } else {
          matches[4].add(hand);
        }
        break;
      case 5:
        matches[0].add(hand);
        break;
      case 6:
        matches[2].add(hand);
        break;
      default:
        print('oh no');
    }
  }

  final convertedMatches = <List<int>>[];

  for (final x in matches) {
    final temp = <List<int>>[];
    for (final i in x) {
      var y = i.cards.replaceAll('A', 'E');
      y = y.replaceAll('T', 'A');
      y = y.replaceAll('J', '1');
      y = y.replaceAll('Q', 'C');
      y = y.replaceAll('K', 'D');
      final val = int.parse(y, radix: 16);
      temp.add([val, i.bid]);
    }
    temp.sort((a, b) => b[0].compareTo(a[0]));
    convertedMatches.addAll(temp);
  }

  int total = 0;
  for (var x = 0; x < convertedMatches.length; x++) {
    total += convertedMatches[x][1] * (convertedMatches.length - x);
  }

  print(total);
}